### PR TITLE
Pro Dashboard: Refactor Notification setting contact form.

### PR DESCRIPTION
**Pull request by jkguidaven**
This PR unifies the SMS and Email editor implementation by adding a shared component that handles both contact type.

Related to 1204774821045518-as-1204802449809227

## Proposed Changes

* Add a Contact Editor component that unifies the implementation for both SMS and Email contact form.
* Split the editor to separate to improve the manageability of the codes.

## Testing Instructions
Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

* Run git checkout add/downtime-monitoring-contact-list-common-component and yarn start-jetpack-cloud
Open http://jetpack.cloud.localhost:3000/, and you'll be redirected to the /dashboard.
Enable monitor if not enabled already.
Click on the clock icon next to the monitor toggle.
Confirm that the email and SMS editor work as expected.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/index.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	const onAddContact = (
+		contact: StateMonitorSettingsEmail | StateMonitorSettingsSMS,
+		asVerified: boolean,
+		sourceEvent?: string
+	) => {
+		if ( sourceEvent ) {
+			recordEvent( sourceEvent );
+		}
+
+		setContacts( addToContactList( type, contacts, contact, asVerified ) );
```
Comment (jkguidaven) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/index.tsx:54]: The logic updating the contact list was separated so we do not have to pass the `setContacts` to the actual form logic.

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx [hunk 2]
```diff
@@ -172,13 +171,17 @@ export default function NotificationSettings( {
 		setHasUnsavedChanges( false );
 	}, [] );
 
-	const handleSetAllEmailItems = ( items: StateMonitorSettingsEmail[] ) => {
-		setAllEmailItems( items );
+	const handleSetAllEmailItems = (
+		items: Array< StateMonitorSettingsEmail | StateMonitorSettingsSMS >
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx]: This should always be `StateMonitorSettingsEmail` only, as it gives the wrong impression.

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx [hunk 3]
```diff
@@ -172,13 +171,17 @@ export default function NotificationSettings( {
...
+	) => {
+		setAllEmailItems( items as Array< StateMonitorSettingsEmail > );
 		setHasUnsavedChanges( false );
 	};
 
-	const handleSetAllPhoneItems = ( items: StateMonitorSettingsSMS[] ) => {
-		setAllPhoneItems( items );
+	const handleSetAllPhoneItems = (
+		items: Array< StateMonitorSettingsEmail | StateMonitorSettingsSMS >
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx]: This should always be `StateMonitorSettingsSMS` only, as it gives the wrong impression.

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+		isError: isRequestingVerificationCodeFailed,
+		isVerified: isRequestingVerificationCodeAlreadyVerified,
+		mutate: requestVerificationCode,
+	} = useRequestVerificationCode();
+
+	const {
+		isLoading: isResendingVerificationCode,
+		isSuccess: isResendingVerificationCodeSuccess,
+		isError: isResendingVerificationCodeFailed,
+		mutate: resendingVerificationCode,
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx]: Can we name this `resendVerificationCode`?

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+	if ( contact && type === 'sms' ) {
+		const { name, countryCode, countryNumericCode, phoneNumber, phoneNumberFull } =
+			contact as StateMonitorSettingsSMS;
+		return {
+			name: name ?? '',
+			countryCode: countryCode ?? '',
+			countryNumericCode: countryNumericCode ?? '',
+			phoneNumber: phoneNumber ?? '',
+			phoneNumberFull: phoneNumberFull ?? '',
+			email: '',
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts]: Do we need `email` for sms?

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+		const { name, email } = contact as StateMonitorSettingsEmail;
+
+		return {
+			name: name ?? '',
+			email: email ?? '',
+			id: email ?? '',
+			countryCode: '',
+			countryNumericCode: '',
+			phoneNumber: '',
+			phoneNumberFull: '',
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts]: Do we need these values for email?

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+		}
+
+		if (
+			type === 'sms' &&
+			( item as StateMonitorSettingsSMS ).phoneNumberFull ===
+				( contact as StateMonitorSettingsSMS ).phoneNumberFull
+		) {
+			isANewContact = false;
+			return {
+				...contacts,
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts]: Shouldn't this be `contact`? Also, I see the return for both SMS and Email is the same. Can we combine these 2?

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/types.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+export type ContactInfo = {
+	id?: string;
+	name: string;
+	email?: string;
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/types.ts:4]: I think it is better to keep SMS & Email contact separate. WDYT?

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+	contactInfo: ContactInfo
+): RequestVerificationCodeParams => {
+	if ( type === 'email' ) {
+		return { type, value: contactInfo.email ?? '', site_ids: [] };
+	}
+
+	if ( type === 'sms' ) {
+		return {
+			type,
+			value: `${ contactInfo.countryNumericCode }${ contactInfo.phoneNumber }`,
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts]: I think we can use `phoneNumberFull` here

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+	type: AllowedMonitorContactTypes,
+	contacts: Array< StateMonitorSettingsEmail | StateMonitorSettingsSMS >,
+	contact: StateMonitorSettingsEmail | StateMonitorSettingsSMS,
+	asVerified: boolean
+) => {
+	let isANewContact = true;
+	const updatedContactList = contacts.map( ( item ) => {
+		if (
+			type === 'email' &&
+			( item as StateMonitorSettingsEmail ).email === ( contact as StateMonitorSettingsEmail ).email
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts]: I see we are having this line in at least 4 places. Do you think it's better to create a common function that executes this line for SMS & Email separately?

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+					name="name"
+					value={ contactInfo.name }
+					disabled={ showCodeVerification }
+					onChange={ handleInputChange( 'name' ) }
+					aria-describedby={ ! isVerifyAction ? 'name-help-text' : undefined }
+				/>
+				{ ! isVerifyAction && (
+					<div className="configure-contact__help-text" id="name-help-text">
+						{ type === 'email' &&
+							translate( 'Give this email a nickname for your personal reference.' ) }
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx]: What do you think about moving the help texts to util? We could create a function(memorized useCallback) to get the help text based on type & input type(name, code)

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx [hunk 12]
```diff
@@ -0,0 +1,509 @@
...
+	// Show error message when email verification fails
+	useEffect( () => {
+		if ( submitVerificationCodeErrorMessage ) {
+			setValidationError( {
+				code: submitVerificationCodeErrorMessage,
+			} );
+		}
+	}, [ translate, submitVerificationCodeErrorMessage ] );
+
+	// Set help text when email verification fails
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx]:
```suggestion
	// Set help text when contact verification fails
```

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+	// Show error message when verification code request fails
+	useEffect( () => {
+		if ( isRequestingVerificationCodeFailed ) {
+			setValidationError( {
+				email: translate( 'Something went wrong. Please try again.' ),
+			} );
+		}
+	}, [ isRequestingVerificationCodeFailed, translate ] );
+
+	// Add email item to the list once the email is verified
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx]:
```suggestion
	// Add email item to the list once the contact is verified
```

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx [hunk 14]
```diff
@@ -0,0 +1,509 @@
...
+		}
+	}, [
+		contactInfo,
+		isRequestingVerificationCodeAlreadyVerified,
+		isSubmittingVerificationCodeSuccess,
+		onAdd,
+		type,
+	] );
+
+	// Show error message when email verification fails
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx]:
```suggestion
	// Show error message when contact verification fails
```

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx [hunk 15]
```diff
@@ -0,0 +1,509 @@
...
+
+		if ( contactInfo?.verificationCode ) {
+			submitVerificationCode( {
+				...getContactInfoPayload( type, contactInfo ),
+				verification_code: Number( contactInfo.verificationCode ),
+			} );
+		}
+	};
+
+	// Add email item to the list if the email is already verified
```
Comment (yashwin) [inline client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx]:
```suggestion
	// Add contact item to the list if the contact is already verified
```

**Review by yashwin — COMMENTED** [review]
I did a quick run of the UI and didn't do thorough testing and it worked as expected. 

I have left a few comments. I'll do another round of thorough testing once the comments are addressed. 

Thanks for working on this.

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx (hunk 2 above)
Comment (jkguidaven) [reply client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx]: Thanks @yashwin for bringing this up. I couldn't figure out how to fix this with minimal change, but this led me to modify the contact editor and utilize Typescript's generic class. 

I ended up cleaning the utils for the contact editor and made significant improvements with the help of generic classes aside the fix for this. 👍

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx (hunk 12 above)
Comment (jkguidaven) [reply client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx]: Good catch. 👍 I fixed the rest of the comments.

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts (hunk 7 above)
Comment (jkguidaven) [reply client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts]: Good catch 👍. This has been combined as suggested.

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts (hunk 10 above)
Comment (jkguidaven) [reply client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/utils.ts]: This has been fixed. 👍

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx (hunk 11 above)
Comment (jkguidaven) [reply client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/verify.tsx]: I agree to move this to a hook. 👍 I also moved `useContactModalTitleAndSubtitle` within Contact editor folder as it is only used within the editor.

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/types.ts (hunk 8 above)
Comment (jkguidaven) [reply client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/types.ts:4]: I initially thought of having a separate contact info class, but this produces a lot of typecasting all over the place, making it more complicated to follow.

If we support another type of contact, we add more typecasting. The solution I found was to have 1 class that could hold different types of contact.

Comment (yashwin) [reply client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-editor/types.ts:4]: Sounds good 👍

File: client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx (hunk 2 above)
Comment (jkguidaven) [reply client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx]: @yashwin I have to backtrack with using the Generic class. It makes the code harder to debug and makes props drilling more complicated. I instead added an alias to improve code readability.

```
export type StateMonitoringSettingsContact = StateMonitorSettingsEmail | StateMonitorSettingsSMS;
```

**Comment by jkguidaven** [comment]
@yashwin, thanks for reviewing my PR. I have addressed all the comments and hope I got everything. Can you have another look?

**Review by yashwin — COMMENTED** [review]
@jkguidaven - Thanks for addressing all the comments.

I faced a couple of issues while testing this.

- Edit a contact doesn't replace the existing one; it instead adds a new one.
- Resend verification endpoint is getting called everything I enter something in the code input field.

**Review by yashwin — DISMISSED** [review]
LGTM 🚀 Tested all the flows, and everything works as expected. 

However, I wasn't able to reproduce the issue with resend. So, I'm going to approve this now. 

Great job on this, @jkguidaven!

**Comment by yashwin** [comment]
Steps to reproduce the issue with resend

Add an email -> Click verify "Later" -> Save the changes -> Try verifying the added email address by clicking on the "Pending" badge -> Enter a text in the code input field and notice the resend API is getting called every time you enter a character.

**Review by vitozev — APPROVED** [review]
As @yashwin already reviewed the PR, I just tested the issues he found and fixed:
- [Issue with resend](https://github.com/Automattic/wp-calypso/pull/78479#issuecomment-1609415915).
- Verify that [resend works for SMS](https://github.com/Automattic/wp-calypso/pull/78720).

These changes are now working as expected.

**Pull request merged**